storage/src/mmr/bitmap.rs

```diff
@@ -171,17 +204,60 @@ impl<H: CHasher> Bitmap<H> {
         // a temporary lightweight (fully pruned) copy of the tree so that we don't require
         // mutability of the original.
         let mut mmr = self.mmr.clone_pruned();
-        let chunk = H::Digest::try_from(self.last_chunk()).unwrap();
-        mmr.add(hasher, &chunk);
+        let last_chunk = self.last_chunk_as_digest();
+        mmr.add(hasher, &last_chunk);
 
         mmr.root(hasher)
     }
+
+    /// Return an inclusion proof for the specified bit, along with the chunk of the bitmap
+    /// containing that bit. The proof can be used to prove any bit in the chunk.
+    pub async fn proof(
+        &self,
+        hasher: &mut H,
+        bit_offset: u64,
+    ) -> Result<(Proof<H>, H::Digest), Error> {
+        assert!(bit_offset < self.bit_count(), "out of bounds");
+
+        let leaf_pos = Self::leaf_pos(bit_offset);
+        let chunk = self.get_chunk_as_digest(bit_offset);
+
+        if self.next_bit == 0 {
+            let proof = Proof::<H>::range_proof(&self.mmr, leaf_pos, leaf_pos).await?;
+            return Ok((proof, chunk));
+        }
+
+        // We must account for the bits in the last chunk.
```

Comment:
Wow. This is clean.

Reply:
yeah I went through many iterations to reduce the ugliness of this boundary condition...